rust/zemu/src/ui_toolkit/backends/nanos.rs

```diff
@@ -0,0 +1,297 @@
+            },
+            Err(_) => {
+                ui.show_error();
+                ui.backend.wait_ui();
+            }
+        }
+    }
+
+    fn wait_ui(&mut self) {
+        unsafe {
+            bindings::crapoline_ux_wait();
+        }
+    }
+
+    fn expert(&self) -> bool {
+        self.expert
+    }
+
+    fn toggle_expert(&mut self) {
+        self.expert = !self.expert;
+
+        self.show_idle(1, None);
+    }
+
+    fn update_expert(&mut self) {
+        let msg = if self.expert {
+            "enabled\x00"
+        } else {
+            "disabled\x00"
+        };
+        let msg = PIC::new(msg).into_inner();
+
+        self.value[..msg.len()].copy_from_slice(msg.as_bytes());
+    }
+
+    fn accept_reject_out(&mut self) -> &mut [u8] {
+        use bolos_sys::raw::G_io_apdu_buffer as APDU_BUFFER;
+
+        unsafe { &mut APDU_BUFFER[..APDU_BUFFER.len() - self.viewable_size] }
+    }
+
+    fn accept_reject_end(&mut self, len: usize) {
+        use bolos_sys::raw::{io_exchange, CHANNEL_APDU, IO_RETURN_AFTER_TX};
+
+        // Safety: simple C call
+        unsafe {
+            io_exchange((CHANNEL_APDU | IO_RETURN_AFTER_TX) as u8, len as u16);
+        }
+    }
+
+    fn store_viewable<V: Viewable + Sized + 'static>(
+        &mut self,
+        viewable: V,
+    ) -> Option<RefMutDynViewable> {
+        use bolos_sys::raw::G_io_apdu_buffer as APDU_BUFFER;
+
+        let size = core::mem::size_of::<V>();
+        unsafe {
+            let buf_len = APDU_BUFFER.len();
+            if size > buf_len {
+                return None;
+            }
+
+            let new_loc_slice = &mut APDU_BUFFER[buf_len - size..];
+            let new_loc_raw_ptr: *mut u8 = new_loc_slice.as_mut_ptr();
+            let new_loc: *mut V = new_loc_raw_ptr.cast();
+
+            //write but we don't want to drop `new_loc` since
+            // it's not actually valid T data
+            core::ptr::write(new_loc, viewable);
+
+            //write how many bytes we have occupied
+            self.viewable_size = size;
+
+            //we can unwrap as we know this ptr is valid
+            Some(new_loc.as_mut().unwrap().into())
+        }
+    }
+}
+
+mod cabi {
+    use super::*;
+
+    #[no_mangle]
+    pub unsafe extern "C" fn rs_h_expert_toggle() {
+        RUST_ZUI.backend.toggle_expert();
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn rs_h_expert_update() {
+        RUST_ZUI.backend.update_expert();
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn rs_h_paging_can_decrease() -> bool {
+        RUST_ZUI.paging_can_decrease()
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn rs_h_paging_can_increase() -> bool {
+        RUST_ZUI.paging_can_increase()
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn view_idle_show_impl(item_idx: u8, status: *mut i8) {
+        let status = if status.is_null() {
+            None
+        } else {
+            let len = crate::ui_toolkit::c_strlen(status as *const u8);
```

Comment:
I am not sure why we need all this.
Could we eventually abstract this whole thing away and move it to rust?
we can probably replace this whole thing with the rust `view_idle` implementation and make the inputs safer?

Reply:
well `view_idle_show` is called outside the ui itself (in main during initialization) so we can't hide this very well unless we plan to remove the arguments themselves from this call...

We could do a "view_idle_show_default" that doesn't take arguments and in that case we could wrap it so it's all in rust

Reply:
ok, let's leave it like this now .. we can review the API once the console+tests exist